pygmt/tests/test_clib_put.py

```diff
@@ -0,0 +1,192 @@
+"""
+Test the functions that put data into GMT.
+"""
+import numpy as np
+import numpy.testing as npt
+import pytest
+
+from .test_clib import mock
+from .. import clib
+from ..exceptions import GMTCLibError, GMTInvalidInput
+from ..helpers import GMTTempFile
+
+
+def test_put_strings():
+    "Check that assigning a numpy array of dtype str to a dataset works"
+    with clib.Session() as lib:
+        dataset = lib.create_data(
+            family="GMT_IS_DATASET|GMT_VIA_VECTOR",
+            geometry="GMT_IS_POINT",
+            mode="GMT_CONTAINER_ONLY",
+            dim=[1, 5, 1, 0],  # columns, rows, layers, dtype
+        )
+        s = np.array(["a", "b", "c", "d", "e"], dtype=np.str)
+        lib.put_strings(dataset, family="GMT_IS_VECTOR", strings=s)
+        # Turns out wesn doesn't matter for Datasets
+        wesn = [0] * 6
+        # Save the data to a file to see if it's being accessed correctly
+        with GMTTempFile() as tmp_file:
+            lib.write_data(
+                "GMT_IS_VECTOR",
+                "GMT_IS_POINT",
+                "GMT_WRITE_SET",
+                wesn,
+                tmp_file.name,
+                dataset,
+            )
+            # Load the data and check that it's correct
+            news = tmp_file.loadtxt(unpack=True, dtype=np.str)
+            npt.assert_allclose(news, s)
```

Comment:
> We can attach the strings to a vector or a matrix. _X_ can be a GMT_VECTOR or GMT_MATRIX, which is created by the `create_data` function. Thus, we have to pass _family_ (`either GMT_IS_VECTOR or GMT_IS_MATRIX`) to let GMT know the type of _X_.

Right, so we can't pass in *just* strings then, but need to tie it to an existing Vector or Matrix? I've tried that (haven't committed it to Github) but am now getting segfaults, which seems promising in a way.

Reply:
I think *dataset* is the Vector or Matrix.

Reply:
I mean, what we want to do is tie the strings to the dataset (the dataset is either a vector or a matrix).

Reply:
Reading this C code (https://github.com/GenericMappingTools/gmt/blob/master/src/testapi_vector_strings.c) may be useful for you.

Reply:
Yes that is helpful, I was reading the GMT.jl one at https://github.com/GenericMappingTools/GMT.jl/blob/master/src/pstext.jl and got really lost.

Reply:
Ah, I think I got it! I was counting the `strings` array as a column using `dim=[4, 5, 1, 0],  # columns, rows, layers, dtype`, but the columns should not be counted. I.e. an array with `x, y, angle, strings` would just use `dim=3` instead of `dim=4`. Hold on, I'll make a commit.

Reply:
PyGMT passes vectors, matrix or grids to GMT. As I understand it, GMT.jl is using a different way passing data to GMT. 

GMT.jl wraps the GMT data types like GMT_DATASET, GMT_GRID, GMT_PALETTE (https://docs.generic-mapping-tools.org/latest/api.html#gmt-resources). So what GMT.jl do is attaching vectors to the GMT_DATASET in Julia, and they pass the GMT_DATASET directly to GMT. This is why we find so many GMT API bugs but everything work for GMT.jl.

Reply:
Ok, maybe not. Using `dim=2` runs without a segfault in 42af00ef8b16afd6c844d648860a6435338dd864, but the 'abcdef ' `text` is not output to the textfile, only the `x` and `y` array data is written. Maybe we do need to use `dim=3`.

Reply:
Actually I'm not sure if it did work, or if it's just that `GMT_Put_Strings` doesn't write the strings to the tmp_file. I'll need to either wrap `GMT_Get_Strings` or refactor `text` to use `put_strings` to be sure. Anyways, dinner first.

Reply:
@weiji14 Here is a minimal example for experimenting with passing strings array to C. I just post it here for your reference. 

The C code (test.c):
```c
#include <stdio.h>

int myfunc (char **array)
{
	fprintf(stderr, "%s %s %s\n", array[0], array[1], array[2]);
	return 0;
}
```

To compile the C code into a shared library, run:
```bash
gcc -fpic -c test.c
gcc -shared -o test.so test.o
```
The following Python code can correctly pass strings array to the myfunc function in test.so library:
```python
import ctypes

lib = ctypes.CDLL("test.so")
print(lib)
string = ["ABC", "DEF", "GHI"]
lib.myfunc.restype = ctypes.c_int

arr = (ctypes.c_char_p * len(string))()
arr[:] = [s.encode() for s in string]
lib.myfunc(arr)
```
The Python trick comes from https://stackoverflow.com/questions/3494598/. It works but I don't understand the details, and don't know if it's the simplest way.